SharkCage/CageServiceInstaller/ServiceInstaller.cs

```diff
@@ -0,0 +1,391 @@
+
+                if (service == IntPtr.Zero)
+                    return false;
+
+                CloseServiceHandle(service);
+                return true;
+            }
+            finally
+            {
+                CloseServiceHandle(scm);
+            }
+        }
+
+        public static void InstallAndStart(string service_name, string display_name, string file_name)
+        {
+            IntPtr scm = OpenSCManager(ScmAccessRights.AllAccess);
+
+            try
+            {
+                IntPtr service = OpenService(scm, service_name, ServiceAccessRights.AllAccess);
+
+                if (service == IntPtr.Zero)
+                    service = CreateService(scm, service_name, display_name, ServiceAccessRights.AllAccess, SERVICE_WIN32_OWN_PROCESS, ServiceBootFlag.AutoStart, ServiceError.Normal, file_name, null, IntPtr.Zero, null, null, null);
+
+                if (service == IntPtr.Zero)
+                    throw new ApplicationException("Failed to install service.");
+
+                try
+                {
+                    StartService(service);
+                }
+                finally
+                {
+                    CloseServiceHandle(service);
+                }
+            }
+            finally
+            {
+                CloseServiceHandle(scm);
+            }
+        }
+
+        public static void StartService(string service_name)
+        {
+            IntPtr scm = OpenSCManager(ScmAccessRights.Connect);
+
+            try
+            {
+                IntPtr service = OpenService(scm, service_name, ServiceAccessRights.QueryStatus | ServiceAccessRights.Start);
+                if (service == IntPtr.Zero)
+                    throw new ApplicationException("Could not open service.");
+
+                try
+                {
+                    StartService(service);
+                }
+                finally
+                {
+                    CloseServiceHandle(service);
+                }
+            }
+            finally
+            {
+                CloseServiceHandle(scm);
+            }
+        }
+
+        public static void StopService(string service_name)
+        {
+            IntPtr scm = OpenSCManager(ScmAccessRights.Connect);
+
+            try
+            {
+                IntPtr service = OpenService(scm, service_name, ServiceAccessRights.QueryStatus | ServiceAccessRights.Stop);
+                if (service == IntPtr.Zero)
+                    throw new ApplicationException("Could not open service.");
+
+                try
+                {
+                    StopService(service);
+                }
+                finally
+                {
+                    CloseServiceHandle(service);
+                }
+            }
+            finally
+            {
+                CloseServiceHandle(scm);
+            }
+        }
+
+        private static void StartService(IntPtr service)
+        {
+            SERVICE_STATUS status = new SERVICE_STATUS();
```

Comment:
what is the status used for?

Reply:
The status is required for the `StartService` dll function. But checking the status right after finishing the function doesn't make sense because of the system needs some time to change the status of the service. Because of that this is handled in the `WaitForServiceStatus` function.

But I optimized the call a bit: `ControlService(service, ServiceControl.Stop, new SERVICE_STATUS());`